new resource "azurerm_iot_security_solution" (#10034)
terraform-providers/terraform-provider-azurerm: website/docs/r/iot_security_solution.html.markdown

}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) Specifies the name of the Iot Security Solution. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) Specifies the name of the resource group in which to create the Iot Security Solution. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.

* `display_name` - (Required) Specifies the Display Name for this Iot Security Solution.

* `iothub_ids` - (Required) Specifies the IoT Hub resource IDs to which this Iot Security Solution is applied.

* `enabled` - (Optional) Is the Iot Security Solution enabled? Defaults to `true`.

* `events_to_export` - (Optional) A list of data which is to exported to analytic workspace. Valid values include `RawEvents`.

* `log_analytics_workspace_id` - (Optional) Specifies the Log Analytics Workspace ID to which the security data will be sent.

* `log_unmasked_ips_enabled` - (Optional) Should ip addressed be unmasked in the log? Defaults to `false`.

> * `recommendations_enabled` - (Optional) A `recommendations_enabled` block as defined below.

Reviewer: ```suggestion
* `recommendations_enabled` - (Optional) A `recommendations_enabled` block of options to enable or disable as defined below.
```

> * `query_for_resources` - (Optional) Azure Resource Graph query which represents the security solution's monitored resources.

Reviewer: grammer
```suggestion
* `query_for_resources` - (Optional) An Azure Resource Graph query used to set the resources monitored.
```

* `query_subscription_ids` - (Optional) A list of subscription Ids on which the user defined resources query should be executed.

* `tags` - (Optional) A mapping of tags to assign to the resource.

---

A `recommendations_enabled` block supports the following:

> * `events_to_exportacr_authentication` - (Optional) Could Service Principal Authentication be used with ACR repository? Defaults to `true`.

Reviewer: We need to fix thes properties:
```suggestion
* `acr_authentication` - (Optional) Is Principal Authentication enabled for the ACR repository? Defaults to `true`.
```

and so forth

Author: oh, sorry for all the typo

}

resource "azurerm_iothub" "example" {
  name                = "example-IoTHub"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location

  sku {
    name     = "S1"
    capacity = "1"
  }
}

resource "azurerm_iot_security_solution" "example" {
  name                = "example-Iot-Security-Solution"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  display_name        = "Iot Security Solution"
  iothub_ids          = [azurerm_iothub.example.id]
}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) Specifies the name of the Iot Security Solution. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) Specifies the name of the resource group in which to create the Iot Security Solution. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.

* `display_name` - (Required) Specifies the Display Name for this Iot Security Solution.

> * `iothub_ids` - (Required) Specifies the IoT Hub resource IDs to which this Iot Security Solution applies.

Reviewer: grammer
```suggestion
* `iothub_ids` - (Required) Specifies the IoT Hub resource IDs to which this Iot Security Solution is applied.
```

* `log_analytics_workspace_id` - (Optional) Specifies the Log Analytics Workspace ID to which the security data will be sent.

* `enabled` - (Optional) Is the Iot Security Solution enabled? Defaults to `true`.

> * `unmasked_ip_logging_enabled` - (Optional) Is unmasked IP logging enabled? Defaults to `false`.

Reviewer: could we reword this?
```suggestion
* `unmasked_ip_logging_enabled` - (Optional) should ip addressed be unmasked in the log? Defaults to `false`.
```

* `iot_ip_filter_permissive_rule_enabled` - (Optional) Is IP filter rule source allowable IP range too large? Defaults to `true`.

* `iot_open_ports_enabled` - (Optional) Is any ports open on the device? Defaults to `true`.

* `iot_permissive_firewall_policy_enabled` - (Optional) Does firewall policy exist which allow necessary communication to/from the device? Defaults to `true`.

* `iot_permissive_input_firewall_rules_enabled` - (Optional) Is only necessary addresses or ports are permitted in? Defaults to `true`.

* `iot_permissive_output_firewall_rules_enabled` - (Optional) Is only necessary addresses or ports are permitted out? Defaults to `true`.

* `iot_privileged_docker_options_enabled` - (Optional) Is high level permissions are needed for the module? Defaults to `true`.

* `iot_shared_credentials_enabled` - (Optional) Is any credentials shared among devices? Defaults to `true`.

* `iot_vulnerable_tls_cipher_suite_enabled` - (Optional) Does TLS cipher suite need to be updated? Defaults to `true`.

---

A `user_defined_resource` block supports the following:

* `query` - (Required) Azure Resource Graph query which represents the security solution's user defined resources.

> * `subscription_ids` - (Required) A list of subscription Ids on which the user defined resources query should be executed..

Reviewer: I think we should improve the ux here, i'm not sure what this is doing exactly? a list of iot hubs to apply this to via a graph query?

Author: here is the doc about `user_defined_resource`: https://docs.microsoft.com/en-us/azure/defender-for-iot/quickstart-configure-your-solution. Through this field, user could monitor the entire IoT solution in one dashboard, surfacing all of your IoT devices, IoT platforms and back-end resources in Azure. 

this is not for auto find iothub. In the portal, the ux is:
![tempsnip](https://user-images.githubusercontent.com/2786738/105442786-5a57e100-5ca5-11eb-8ff1-b617aaf60c0c.png)

user choose the sub or resourceGroup, the portal will convert into graph query and send to the backend. Should we do like the portal and dynamically build the graph?

Reviewer: Maybe it would be best to rename these to query_for_resources &  query_subscription_ids?

Author: updated
